src/server/zset_family.cc

```diff
@@ -691,23 +690,39 @@ void InterScoredMap(ScoredMap* dest, ScoredMap* src, AggType agg_type) {
     dest->swap(*src);
 }
 
-OpResult<ScoredMap> OpUnion(EngineShard* shard, Transaction* t, string_view dest, AggType agg_type,
-                            const vector<double>& weights, bool store) {
-  ArgSlice keys = t->ShardArgsInShard(shard->shard_id());
-  DVLOG(1) << "shard:" << shard->shard_id() << ", keys " << vector(keys.begin(), keys.end());
-  DCHECK(!keys.empty());
-
-  unsigned start = 0;
+template <typename SliceIt>
+ScoredMap UnionSliceResultScore(SliceIt start, SliceIt end, AggType agg_type) {
+  ScoredMap result;
+  for (auto it = start; it != end; ++it) {
+    if (it->first.is_done()) {
+      continue;
+    }
 
-  if (keys.front() == dest) {
-    ++start;
+    ScoredMap sm = FromObject(it->first->second, it->second);
+    if (result.empty()) {
+      result.swap(sm);
+    } else {
+      UnionScoredMap(&result, &sm, agg_type);
+    }
   }
+  return result;
+}
+
+using ScoredPI = pair<PrimeIterator, double>;
+using ScoredPIVec = vector<ScoredPI>;
+
+double SliceResultWeight(EngineShard* shard, Transaction* t, const vector<double>& weights,
+                         unsigned key_index, unsigned offset) {
+  unsigned windex = t->ReverseArgIndex(shard->shard_id(), key_index) - offset;
+  DCHECK_LT(windex, weights.size());
+  return weights[windex];
+}
 
+OpResult<ScoredPIVec> FindSliceResult(EngineShard* shard, Transaction* t, const ArgSlice& keys,
```

Comment:
ArgSlice type is absl::Span, better pass by value and not const &

Reply:
In this function you are retrieving the slice items and corresponding weight, I would rename the function so it would describe this functionality i.e FindSliceKeysAndWeights